### PR TITLE
SolutionSubmission 2nd step: undoing previous solution and reapplying new solutions

### PR DESCRIPTION
**Pull request by josojo**
Implements:
undoing the previous solution and reapplying new solutions, if there was a previous solution submitted for the same batch.

Testplan:
uint tests only..

### CODE REVIEW

File: contracts/StablecoinConverter.sol [hunk 1]
```diff
@@ -167,4 +174,49 @@ contract StablecoinConverter is EpochTokenLocker {
...
+                    previousSolutionPrices[findPriceIndex(order.sellToken, previousSolutionTokenIdForPrice)];
+
+                order.buyAmount = uint128(order.sellAmount.add(sellVolume).mul(order.buyAmount) / order.sellAmount);
+                order.sellAmount = uint128(order.sellAmount.add(sellVolume));
+                orders[previousSolutionTrades[i].owner][previousSolutionTrades[i].orderId] = order;
+                subtractBalance(previousSolutionTrades[i].owner, tokenIdToAddressMap(order.buyToken), buyVolume);
+                addBalance(previousSolutionTrades[i].owner, tokenIdToAddressMap(order.sellToken), sellVolume);
+            }
+        }
+        delete previousSolutionTrades;
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: I'd maybe reset any existing state in the function that writes it, otherwise you have an implicit call order dependency between these two functions. The method that needs these fields to be empty should make sure they are empty.

File: contracts/StablecoinConverter.sol [hunk 2]
```diff
@@ -167,4 +174,49 @@ contract StablecoinConverter is EpochTokenLocker {
...
+                Order memory order = orders[previousSolutionTrades[i].owner][previousSolutionTrades[i].orderId];
+                uint sellVolume = previousSolutionTrades[i].volume;
+                uint buyVolume = previousSolutionTrades[i].volume
+                    .mul(previousSolutionPrices[findPriceIndex(order.buyToken, previousSolutionTokenIdForPrice)]) /
+                    previousSolutionPrices[findPriceIndex(order.sellToken, previousSolutionTokenIdForPrice)];
+
+                order.buyAmount = uint128(order.sellAmount.add(sellVolume).mul(order.buyAmount) / order.sellAmount);
+                order.sellAmount = uint128(order.sellAmount.add(sellVolume));
+                orders[previousSolutionTrades[i].owner][previousSolutionTrades[i].orderId] = order;
+                subtractBalance(previousSolutionTrades[i].owner, tokenIdToAddressMap(order.buyToken), buyVolume);
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: I think you might have to do the subtractions after the additions again since otherwise you might become temporarily negative. A unit test for this would be nice.

File: contracts/StablecoinConverter.sol [hunk 3]
```diff
@@ -167,4 +174,49 @@ contract StablecoinConverter is EpochTokenLocker {
...
+        for (uint i = 0; i < len; i++) {
+            previousSolutionPrices.push(prices[i]);
+            previousSolutionTokenIdForPrice.push(tokenIdForPrice[i]);
+        }
+    }
+
+    function undoPreviousSolution(uint batchIndex) internal {
+        if (previousSolutionBatchId == batchIndex) {
+            for (uint i = 0; i < previousSolutionTrades.length; i++) {
+                Order memory order = orders[previousSolutionTrades[i].owner][previousSolutionTrades[i].orderId];
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Can we store `previousSolutionTrades[i]` in a local variable for readability? This code is very hard to read.

File: contracts/StablecoinConverter.sol [hunk 4]
```diff
@@ -167,4 +174,49 @@ contract StablecoinConverter is EpochTokenLocker {
...
+        uint batchIndex,
+        address[] memory owner,  //tradeData is submitted as arrays
+        uint16[] memory orderId,
+        uint128[] memory volume,
+        uint128[] memory prices,
+        uint16[] memory tokenIdForPrice
+    ) internal {
+        previousSolutionBatchId = batchIndex;
+        uint len = owner.length;
+        for (uint i = 0; i < len; i++) {
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: nit: why this len assignment? It's only used once.

File: test/stablecoin_converter.js [hunk 5]
```diff
@@ -223,6 +223,147 @@ contract("StablecoinConverter", async (accounts) => {
...
+
+      await stablecoinConverter.submitSolution(batchIndex, owner, orderId, volume3, prices, tokenIdForPrice)
+      
+      assert.equal((await stablecoinConverter.getBalance.call(user_1, erc20_1.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_1, erc20_2.address), 20, "Bought tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_2, erc20_1.address), 10, "Bought tokens were not adjusted correctly")
+      assert.equal((await stablecoinConverter.getBalance.call(user_2, erc20_2.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+    
+    })
+    it("checks that solutiontrades are delete from previous batchIndex", async () => {
```
Comment (fleupold) [inline test/stablecoin_converter.js]: typo: deleted

I don't really understand what you are testing here? That a batch is correctly reverted? Maybe you can rephrase to make it more clear.

File: test/stablecoin_converter.js [hunk 6]
```diff
@@ -223,6 +223,147 @@ contract("StablecoinConverter", async (accounts) => {
...
+      await stablecoinConverter.addToken(erc20_2.address)
+      const batchIndex = (await stablecoinConverter.getCurrentStateIndex.call()).toNumber()
+
+      const orderId1 = await sendTxAndGetReturnValue(stablecoinConverter.placeOrder, 1, 0, true, batchIndex + 1, 20, 10, {from: user_1})
+      const orderId2 = await sendTxAndGetReturnValue(stablecoinConverter.placeOrder, 0, 1, true, batchIndex + 1, 10, 20, {from: user_2})
+      // close auction
+      await waitForNSeconds(BATCH_TIME)
+
+      const prices = [10, 20]
+      const owner = [user_1, user_2]  //tradeData is submitted as arrays
```
Comment (fleupold) [inline test/stablecoin_converter.js:247]: Having to write this all the time is a sign that really tradeData should be an object and we should pass an array of those.

File: contracts/StablecoinConverter.sol (hunk 2 above)
Comment (bh2smith) [reply contracts/StablecoinConverter.sol]: I  ❤️  unit tests!

**Review by bh2smith — APPROVED** [review]
I am in favour of the existing comments for improvement here.

File: contracts/StablecoinConverter.sol (hunk 4 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: I did the gas test. it saved 11 gas :D 

You are right. I will change it.

**Review by bh2smith — APPROVED** [review]
This is starting to look really good nice extensive testing!

File: contracts/StablecoinConverter.sol [hunk 7]
```diff
@@ -110,12 +110,15 @@ contract StablecoinConverter is EpochTokenLocker {
         return IdToAddressBiMap.getAddressAt(registeredTokens, id);
     }
 
+    uint public previousSolutionBatchId;
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Would it make sense to group these on a previousSolution struct?

File: contracts/StablecoinConverter.sol [hunk 8]
```diff
@@ -173,12 +182,62 @@ contract StablecoinConverter is EpochTokenLocker {
...
+    ) internal {
+        previousSolutionBatchId = batchIndex;
+        for (uint i = 0; i < owners.length; i++) {
+            previousSolutionTrades.push(TradeData({
+                owner: owners[i],
+                orderId: orderIds[i],
+                volume: volumes[i]
+            }));
+        }
+        for (uint i = 0; i < prices.length; i++) {
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: nit: should iterate over the array you are accessing. `previousSolutionTrades.length` instead of `prices.length`

File: contracts/StablecoinConverter.sol [hunk 9]
```diff
@@ -173,12 +182,62 @@ contract StablecoinConverter is EpochTokenLocker {
...
+            previousSolutionTokenIdForPrice.push(tokenIdsForPrice[i]);
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Can you not assign the array directly?

`previousSolutionTokenIdForPrice = tokenIdsForPrice`

File: contracts/StablecoinConverter.sol [hunk 10]
```diff
@@ -173,12 +182,62 @@ contract StablecoinConverter is EpochTokenLocker {
...
+        }
+    }
+
+    function undoPreviousSolution(uint batchIndex) internal {
+        if (previousSolutionBatchId == batchIndex) {
+            for (uint i = 0; i < previousSolutionTrades.length; i++) {
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: What are we going to do in order to prevent this from running out of gas and thus rendering the previous solution as un-revertable?

File: contracts/StablecoinConverter.sol [hunk 11]
```diff
@@ -173,12 +182,62 @@ contract StablecoinConverter is EpochTokenLocker {
...
+                Order memory order = orders[owner][orderId];
+                uint sellVolume = previousSolutionTrades[i].volume;
+                addBalance(owner, tokenIdToAddressMap(order.sellToken), sellVolume);
+            }
+            for (uint i = 0; i < previousSolutionTrades.length; i++) {
+                address owner = previousSolutionTrades[i].owner;
+                uint orderId = previousSolutionTrades[i].orderId;
+                Order memory order = orders[owner][orderId];
+                uint sellVolume = previousSolutionTrades[i].volume;
+                uint buyVolume = sellVolume
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Does this logic already exist somewhere else and can be refactored in a common method?

File: test/stablecoin_converter.js [hunk 12]
```diff
@@ -223,6 +223,186 @@ contract("StablecoinConverter", async (accounts) => {
       assert.equal((orderResult2.sellAmount).toNumber(), 10, "sellAmount was stored incorrectly")
       assert.equal((orderResult2.buyAmount).toNumber(), 5, "buyAmount was stored incorrectly")
     })
+    it("places two orders and first matches them partically and then fully in a 2nd solution submission", async () => {
```
Comment (fleupold) [inline test/stablecoin_converter.js:226]: partically -> partially

Comment (fleupold) [reply test/stablecoin_converter.js:226]: nit: maybe rephrase into "Accepts a second solution that achieves more volume than existing one"

File: test/stablecoin_converter.js [hunk 13]
```diff
@@ -223,6 +223,186 @@ contract("StablecoinConverter", async (accounts) => {
...
+      const volume3 = [10, 20]
+
+      await stablecoinConverter.submitSolution(batchIndex, owner, orderId, volume3, prices, tokenIdForPrice)
+
+      assert.equal((await stablecoinConverter.getBalance.call(user_1, erc20_1.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_1, erc20_2.address), 20, "Bought tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_2, erc20_1.address), 10, "Bought tokens were not adjusted correctly")
+      assert.equal((await stablecoinConverter.getBalance.call(user_2, erc20_2.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+    })
+    it("checks that solutiontrades are delete even if balances get temporarly negative while reverting ", async () => {
```
Comment (fleupold) [inline test/stablecoin_converter.js]: solution trades, deleted, temporarily

File: test/stablecoin_converter.js [hunk 14]
```diff
@@ -223,6 +223,186 @@ contract("StablecoinConverter", async (accounts) => {
...
+      assert.equal(await stablecoinConverter.getBalance.call(user_1, erc20_2.address), 20, "Bought tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_2, erc20_1.address), 10, "Bought tokens were not adjusted correctly")
+      assert.equal((await stablecoinConverter.getBalance.call(user_2, erc20_2.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+    })
+    it("checks solution submission with 3 different solutions", async () => {
+      const stablecoinConverter = await StablecoinConverter.new(2 ** 16 - 1)
+      const erc20_1 = await MockContract.new()
+      const erc20_2 = await MockContract.new()
+
+      await erc20_1.givenAnyReturnBool(true)
```
Comment (fleupold) [inline test/stablecoin_converter.js:274]: I'd be ok with having a utility method that registers n tokens (since it's used in all tests).

File: test/stablecoin_converter.js [hunk 15]
```diff
@@ -223,6 +223,186 @@ contract("StablecoinConverter", async (accounts) => {
...
+
+      await stablecoinConverter.submitSolution(batchIndex, owner, orderId, volume2, prices, tokenIdForPrice)
+
+      assert.equal((await stablecoinConverter.getBalance.call(user_1, erc20_1.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_1, erc20_2.address), 20, "Bought tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_2, erc20_1.address), 10, "Bought tokens were not adjusted correctly")
+      assert.equal((await stablecoinConverter.getBalance.call(user_2, erc20_2.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+    })
+    it("checks solution submission with 3 different solutions", async () => {
+      const stablecoinConverter = await StablecoinConverter.new(2 ** 16 - 1)
```
Comment (fleupold) [inline test/stablecoin_converter.js:270]: We cannot have two constructors (one that takes a manual value, one that takes a concrete)?

File: test/stablecoin_converter.js [hunk 16]
```diff
@@ -223,6 +223,186 @@ contract("StablecoinConverter", async (accounts) => {
...
+      const volume2 = [10, 20]
+
+      await stablecoinConverter.submitSolution(batchIndex, owner, orderId, volume2, prices, tokenIdForPrice)
+
+      assert.equal((await stablecoinConverter.getBalance.call(user_1, erc20_1.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_1, erc20_2.address), 20, "Bought tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_2, erc20_1.address), 10, "Bought tokens were not adjusted correctly")
+      assert.equal((await stablecoinConverter.getBalance.call(user_2, erc20_2.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+    })
+    it("checks solution submission with 3 different solutions", async () => {
```
Comment (fleupold) [inline test/stablecoin_converter.js]: What's the point of this test? Maybe this text could be more clear.

File: test/stablecoin_converter.js [hunk 17]
```diff
@@ -223,6 +223,186 @@ contract("StablecoinConverter", async (accounts) => {
...
+      assert.equal((await stablecoinConverter.getBalance.call(user_1, erc20_1.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+      assert.equal((await stablecoinConverter.getBalance.call(user_2, erc20_1.address)).toNumber(), 0, "Sold tokens were not adjusted correctly")
+      assert.equal((await stablecoinConverter.getBalance.call(user_3, erc20_1.address)).toNumber(), 10, "Sold tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_1, erc20_2.address), 10, "Bought tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_2, erc20_2.address), 0, "Bought tokens were not adjusted correctly")
+      assert.equal(await stablecoinConverter.getBalance.call(user_3, erc20_2.address), 0, "Bought tokens were not adjusted correctly")
+      await stablecoinConverter.submitSolution(batchIndex, owner, orderId, volume, prices, tokenIdsForPrice)
+
+    })
+    it("checks that trades documented from a previous trade are deleted, before new trades are documented", async () => {
```
Comment (fleupold) [inline test/stablecoin_converter.js]: Isn't this already tested by the basic override example? Or do you mean trades from a previous batch?

**Review by fleupold — APPROVED** [review]
Looks good overall. Some suggestions inline. I think the unit test descriptions could be improved.

<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia3.giphy.com%2Fmedia%2Fl2JJrEx9aRsjNruhi%2Fgiphy.gif&token=looks_good" data-gitmeme-token="looks_good"></a>

File: contracts/StablecoinConverter.sol (hunk 8 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: good catch.

File: contracts/StablecoinConverter.sol (hunk 10 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: Let's tackle this later:
https://github.com/gnosis/dex-contracts/issues/171

File: contracts/StablecoinConverter.sol (hunk 11 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: I know you ask these questions rhetorically :D
However, I currently see only the calculation for sellVolume and buyVolume as refactorable. But is it worth doing so, if these are just one single line each?

File: test/stablecoin_converter.js (hunk 13 above)
Comment (josojo) [reply test/stablecoin_converter.js]: I simply don't see these things! shit. I will try again to install a spell checker for unit tests

File: test/stablecoin_converter.js (hunk 15 above)
Comment (josojo) [reply test/stablecoin_converter.js:270]: Yeah, it gives us the following error:
DeclarationError: More than one constructor defined

File: test/stablecoin_converter.js (hunk 14 above)
Comment (josojo) [reply test/stablecoin_converter.js:274]: let me do this in the next PR. I have already rewritten it in there with a fee token and the merge would be quite complicated, but yeah, good comment

File: contracts/StablecoinConverter.sol (hunk 7 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: yeah...

File: contracts/StablecoinConverter.sol (hunk 11 above)
Comment (fleupold) [reply contracts/StablecoinConverter.sol]: I think any complex logic (even if it's just one line) can be refactored in a method. It's about having this logic in one place in case we need to change it later.

File: contracts/StablecoinConverter.sol [hunk 18]
```diff
@@ -173,12 +198,57 @@ contract StablecoinConverter is EpochTokenLocker {
...
+                owner: owners[i],
+                orderId: orderIds[i],
+                volume: volumes[i]
+            }));
+        }
+        previousSolution.tokenIdsForPrice = tokenIdsForPrice;
+    }
+
+    function undoPreviousSolution(uint batchIndex) internal {
+        if (previousSolution.batchId == batchIndex) {
```
Comment (anxolin) [inline contracts/StablecoinConverter.sol:229]: nit: I think this `if` would be better in the main function, not in the internal one. 

```solidity
// Unapply any previous solution
if (previousSolution.batchId == batchIndex) {
    undoPreviousSolution(batchIndex);
}
```

File: contracts/StablecoinConverter.sol [hunk 19]
```diff
@@ -173,12 +198,57 @@ contract StablecoinConverter is EpochTokenLocker {
...
+            for (uint i = 0; i < previousSolution.trades.length; i++) {
```
Comment (anxolin) [inline contracts/StablecoinConverter.sol:230]: nit: It would really help adding a summary on top of each for loop, just summarising what it does
```solidity
// Restore previous balances for the sellers
for (uint i = 0; i < previousSolution.trades.length; i++) {
    //...
}
```

File: contracts/StablecoinConverter.sol [hunk 20]
```diff
@@ -173,12 +198,57 @@ contract StablecoinConverter is EpochTokenLocker {
...
+                address owner = previousSolution.trades[i].owner;
+                uint orderId = previousSolution.trades[i].orderId;
+                Order memory order = orders[owner][orderId];
+                uint sellVolume = previousSolution.trades[i].volume;
+                addBalance(owner, tokenIdToAddressMap(order.sellToken), sellVolume);
+            }
+            for (uint i = 0; i < previousSolution.trades.length; i++) {
```
Comment (anxolin) [inline contracts/StablecoinConverter.sol:237]: We first restore the sellers balance and then the buyers so we prevent negative amounts in case an account did some sells and buys?

**Review by anxolin — COMMENTED** [review]
Super good one!

I would approve but it's merged. Sorry for the late review.
Added some super low importance nits as comments.

**Review by bh2smith — COMMENTED** [review]
Also too late for me to review, but looks pretty good. No point in making suggestions at this time!
